lib/juttle-service-client.js

```diff
@@ -0,0 +1,347 @@
+    },
+    'wait': {
+        usage: '--wait',
+        desc: 'If true, wait for program to finish, otherwise run in background'
+    },
+    'input': {
+        usage: '--input name=val',
+        desc: 'One or more input values.'
+    },
+    'job': {
+        usage: '--job <job-id>',
+        desc: 'Job id.'
+    },
+    'observer': {
+        usage: '--observer <observer-id>',
+        desc: 'Observer id.'
+    },
+    'server': {
+        usage: '--juttle-service <hostname:port>',
+        desc: 'Hostname/port of server'
+    },
+    help: {
+        usage: '--help',
+        desc: 'Print this help and exit'
+    }
+}
+
+let commands = {
+    subscribe: {
+        usage: '({job} | {observer})',
+        perform: subscribe
+    },
+    list_jobs: {
+        usage: '[{job}]',
+        perform: list_jobs
+    },
+    list_observers: {
+        usage: '',
+        perform: list_observers
+    },
+    run: {
+        usage: '{path} [{wait}] [{observer}]',
+        perform: run
+    },
+    delete: {
+        usage: '{job}',
+        perform: delete_job
+    },
+    get_inputs: {
+        usage: '{path} {input} [{input} ...]',
+        perform: get_inputs
+    }
+};
+
+function expand(string) {
+    return string.replace(/\{([^}]*)\}/g, function(match, key) {
+        return cmdargs[key].usage;
+    });
+}
+
+function init(options) {
+    write_output = options.write_output;
+    write_error = options.write_error;
+    client_exit = options.client_exit;
+    commands = _.extend(commands, options.commands);
+    cmdargs = _.extend(cmdargs, options.cmdargs);
+}
+
+function usage() {
+
+    // Given the command descriptions, create the list of commands
+    // used by each option.
+    _.each(cmdargs, function (arginfo, arg) {
+        arginfo.commands = [];
+        _.each(commands, function(info, cmd) {
+            if (info.usage.indexOf('{' + arg + '}') > -1) {
+                arginfo.commands.push(cmd);
+            }
+        });
+    });
+
+    // Expand each command usage
+    _.each(commands, function (info, cmd) {
+        info.usage = expand(info.usage);
+    });
+
+    write_output(sprintf('usage: [%s] [--help] [COMMAND] [OPTIONS]', cmdargs['server'].usage));
+    write_output(sprintf('%5s[COMMAND]: one of the following, with the following options:', ''));
+    _.each(commands, function (info, cmd) {
+        write_output(sprintf('%10s%s %s', '', cmd, info.usage));
+    });
+    write_output(sprintf('%5s[OPTIONS]: one of the following:', ''));
+    _.each(cmdargs, function (info, arg) {
+        write_output(sprintf('%10s%-45s%s','', info.usage, info.desc));
+        if (info.commands.length > 0) {
+            write_output(sprintf('%10s%-45sused by: %s', '', '', info.commands.join(',')));
+        }
+    });
+    client_exit(1);
+}
+
+function command(server, opts, command) {
+    http_url = 'http://' + server + API_PREFIX;
+    ws_url = 'ws://' + server + API_PREFIX;
```

Comment:
Seems better to stash these in opts and pass them through to the command that way rather than in globals.